### DISCUSSION TITLE
Custom Mapping not Working in cmdline Mode

### DISCUSSION THREAD

**Discussion by Balrrach** [hrsh7th/nvim-cmp#1340] [Q&A]
These are my mapping in cmp.setup:
```vim
mapping = cmp.mapping.preset.insert({
	['<C-d>'] = cmp.mapping.scroll_docs(-4),
	['<C-f>'] = cmp.mapping.scroll_docs(4),
	['<C-Space>'] = cmp.mapping.complete(),
	['<CR>'] = cmp.mapping.confirm({ select = true }),
	['<C-n>'] = cmp.mapping(cmp.mapping.select_next_item(), { 'i', 's' }),
	['<C-e>'] = cmp.mapping(cmp.mapping.select_prev_item(), { 'i', 's' }),
	['<Esc>'] =
	function(fallback)
		cmp.mapping.abort()
		fallback()
	end,
    }),
```

This is my configuration for cmdline:
```vim
cmp.setup.cmdline(':', {
    mapping = cmp.mapping.preset.cmdline(),

    sources = cmp.config.sources({
      { name = 'path' }
    }, {
      { name = 'cmdline' }
    })
  })

```

In my case `ctrl+e` is mapped to `select_previous_item()` and this works correctly usually but not but not in  `cmdline` mode. In `cmdline` instead of selecting previous item suggestion is cancelled as if no mapping was made at all since this is the standard behavior. I can't guess the cause since other mappings like `ctrl+n` to `select_next_item()` works correctly. Before posting it as an issue I'd like to see if anyone can give a hint on the cause of this strange behavior. Thanks before hand!

**Comment by jbertozzi** [comment]
I ran into the same issue, I couldn't make it work. :(

**Comment by fitrh** [comment]
If you want full control over the keymap, don't use the preset

```lua
mapping = {
    -- add 'c' to make it work in cmdline
    ['<C-e>'] = cmp.mapping(cmp.mapping.select_prev_item(), { 'i', 's', 'c' }),
    -- or this form
    ['<C-e>'] = {
        i = cmp.mapping.select_prev_item(),
        c = cmp.mapping.select_prev_item(),
    },
   -- also, inside a closure, you can't use `cmp.mapping` helper, instead, use cmp function directly
   ['<Esc>'] = function(fallback)
        cmp.abort() -- instead of cmp.mapping.abort()
        fallback()
    end,
}
```

You can see what that `cmp.preset` doing in here: https://github.com/hrsh7th/nvim-cmp/blob/main/lua/cmp/config/mapping.lua
